Flaky test: UserSessionConcurrencyTest.testConcurrentNotesChange

### Description

```
From 2022-12-13 to 2022-12-19
============================================================================================================================================
Finding details for UserSessionConcurrencyTest.testConcurrentNotesChange
============================================================================================================================================
https://github.com/keycloak/keycloak/actions/runs/3728587458

Store Model Tests	Run model tests	2022-12-19T05:44:48.7315298Z [ERROR]   UserSessionConcurrencyTest.testConcurrentNotesChange:88->lambda$testConcurrentNotesChange$4:88->KeycloakModelTest.inComittedTransaction:540->KeycloakModelTest.inComittedTransaction:553 » Model javax.persistence.PersistenceException: org.hibernate.TransactionException: Unable to commit against JDBC Connection
--------------------------------------------------------------------------------------------------------------------------------------------
https://github.com/keycloak/keycloak/actions/runs/3726993137

Store Model Tests	Run model tests	2022-12-18T23:48:21.2992544Z [ERROR]   UserSessionConcurrencyTest.testConcurrentNotesChange:88->lambda$testConcurrentNotesChange$4:88->KeycloakModelTest.inComittedTransaction:540->KeycloakModelTest.inComittedTransaction:553 » Model javax.persistence.PersistenceException: org.hibernate.TransactionException: Unable to commit against JDBC Connection
--------------------------------------------------------------------------------------------------------------------------------------------
https://github.com/keycloak/keycloak/actions/runs/3726378057

Store Model Tests	Run model tests	2022-12-18T20:53:36.7738501Z [ERROR]   UserSessionConcurrencyTest.testConcurrentNotesChange:88 » Model org.keycloak.models.ModelException: javax.persistence.PersistenceException: org.hibernate.TransactionException: Unable to commit against JDBC Connection
--------------------------------------------------------------------------------------------------------------------------------------------
```

```
============================================================================================================================================
Finding details for UserSessionConcurrencyTest.testConcurrentNotesChange
============================================================================================================================================
https://github.com/keycloak/keycloak/actions/runs/3734776037

Store Model Tests	Run model tests	2022-12-19T20:47:32.3333630Z [ERROR]   UserSessionConcurrencyTest.testConcurrentNotesChange:88 » Model org.keycloak.models.ModelException: javax.persistence.PersistenceException: org.hibernate.TransactionException: Unable to commit against JDBC Connection
--------------------------------------------------------------------------------------------------------------------------------------------
https://github.com/keycloak/keycloak/actions/runs/3728587458

Store Model Tests	Run model tests	2022-12-19T05:44:48.7315298Z [ERROR]   UserSessionConcurrencyTest.testConcurrentNotesChange:88->lambda$testConcurrentNotesChange$4:88->KeycloakModelTest.inComittedTransaction:540->KeycloakModelTest.inComittedTransaction:553 » Model javax.persistence.PersistenceException: org.hibernate.TransactionException: Unable to commit against JDBC Connection
--------------------------------------------------------------------------------------------------------------------------------------------
```

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientModelTest.java
@@ -59,7 +59,7 @@ public class ClientModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientScopeStorageTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientScopeStorageTest.java
@@ -40,7 +40,7 @@ public class ClientScopeStorageTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/ConcurrentHashMapStorageTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/ConcurrentHashMapStorageTest.java
@@ -68,7 +68,7 @@ public class ConcurrentHashMapStorageTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
@@ -118,6 +118,7 @@ import org.keycloak.models.DeploymentStateProviderFactory;
  * @author hmlnarik
  */
 public abstract class KeycloakModelTest {
+    public static final String KEYCLOAK_MODELTESTS_RETRY_TRANSACTIONS = "keycloak.modeltests.retry-transactions";
 
     private static final Logger LOG = Logger.getLogger(KeycloakModelParameters.class);
     private static final AtomicInteger FACTORY_COUNT = new AtomicInteger();
@@ -549,22 +550,37 @@ public abstract class KeycloakModelTest {
     }
 
     protected <T, R> R inComittedTransaction(T parameter, BiFunction<KeycloakSession, T, R> what, BiConsumer<KeycloakSession, T> onCommit, BiConsumer<KeycloakSession, T> onRollback) {
-        AtomicReference<R> res = new AtomicReference<>();
-        KeycloakModelUtils.runJobInTransaction(getFactory(), session -> {
-            session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
-                @Override
-                protected void commitImpl() {
-                    if (onCommit != null) { onCommit.accept(session, parameter); }
-                }
+        if (Boolean.parseBoolean(System.getProperty(KEYCLOAK_MODELTESTS_RETRY_TRANSACTIONS, "false"))) {
+            return KeycloakModelUtils.runJobInRetriableTransaction(getFactory(), session -> {
+                session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
+                    @Override
+                    protected void commitImpl() {
+                        if (onCommit != null) { onCommit.accept(session, parameter); }
+                    }
 
-                @Override
-                protected void rollbackImpl() {
-                    if (onRollback != null) { onRollback.accept(session, parameter); }
-                }
+                    @Override
+                    protected void rollbackImpl() {
+                        if (onRollback != null) { onRollback.accept(session, parameter); }
+                    }
+                });
+                return what.apply(session, parameter);
+            }, 5, 100);
+        } else {
+            return KeycloakModelUtils.runJobInTransactionWithResult(getFactory(), session -> {
+                session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
+                    @Override
+                    protected void commitImpl() {
+                        if (onCommit != null) { onCommit.accept(session, parameter); }
+                    }
+
+                    @Override
+                    protected void rollbackImpl() {
+                        if (onRollback != null) { onRollback.accept(session, parameter); }
+                    }
+                });
+                return what.apply(session, parameter);
             });
-            res.set(what.apply(session, parameter));
-        });
-        return res.get();
+        }
     }
 
     /**
@@ -591,6 +607,16 @@ public abstract class KeycloakModelTest {
             Thread.currentThread().interrupt();
             throw new RuntimeException(ex);
         }
+    }
+
+    protected static RealmModel createRealm(KeycloakSession s, String name) {
+        RealmModel realm = s.realms().getRealmByName(name);
+        if (realm != null) {
+            // The previous test didn't clean up the realm for some reason, cleanup now
+            s.realms().removeRealm(realm.getId());
+        }
+        realm = s.realms().createRealm(name);
+        return realm;
     }
 
 }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/MigrationModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/MigrationModelTest.java
@@ -44,7 +44,7 @@ public class MigrationModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/RealmModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/RealmModelTest.java
@@ -50,7 +50,7 @@ public class RealmModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java
@@ -78,7 +78,7 @@ public class UserModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
 

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserPaginationTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserPaginationTest.java
@@ -40,7 +40,7 @@ public class UserPaginationTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
 

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/authz/ConcurrentAuthzTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/authz/ConcurrentAuthzTest.java
@@ -61,7 +61,7 @@ public class ConcurrentAuthzTest extends KeycloakModelTest {
 
     @Override
     protected void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
 
         realmId = realm.getId();

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
@@ -47,7 +47,7 @@ public class AdminEventQueryTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/EventQueryTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/EventQueryTest.java
@@ -49,7 +49,7 @@ public class EventQueryTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/group/GroupModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/group/GroupModelTest.java
@@ -36,7 +36,7 @@ public class GroupModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorageCockroachdb.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorageCockroachdb.java
@@ -50,6 +50,7 @@ import org.keycloak.provider.Spi;
 import org.keycloak.sessions.AuthenticationSessionSpi;
 import org.keycloak.testsuite.model.Config;
 import org.keycloak.testsuite.model.KeycloakModelParameters;
+import org.keycloak.testsuite.model.KeycloakModelTest;
 import org.testcontainers.containers.CockroachContainer;
 import org.testcontainers.utility.DockerImageName;
 
@@ -128,6 +129,7 @@ public class JpaMapStorageCockroachdb extends KeycloakModelParameters {
 
             COCKROACHDB_DB_JDBC_URL = COCKROACHDB_CONTAINER.getJdbcUrl();
         }
+        System.setProperty(KeycloakModelTest.KEYCLOAK_MODELTESTS_RETRY_TRANSACTIONS, "true");
     }
 
     @Override

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/role/RoleModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/role/RoleModelTest.java
@@ -39,7 +39,7 @@ public class RoleModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         this.realmId = realm.getId();
 

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/AuthenticationSessionTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/AuthenticationSessionTest.java
@@ -49,7 +49,7 @@ public class AuthenticationSessionTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setAccessCodeLifespanLogin(1800);
 

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java
@@ -50,7 +50,7 @@ public class HotRodUserSessionClientSessionRelationshipTest extends KeycloakMode
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setSsoSessionIdleTimeout(1800);
         realm.setSsoSessionMaxLifespan(36000);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java
@@ -85,7 +85,7 @@ public class OfflineSessionPersistenceTest extends KeycloakModelTest {
     }
 
     private static RealmModel prepareRealm(KeycloakSession s, String name) {
-        RealmModel realm = s.realms().createRealm(name);
+        RealmModel realm = createRealm(s, name);
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setSsoSessionMaxLifespan(10 * 60 * 60);
         realm.setSsoSessionIdleTimeout(1 * 60 * 60);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java
@@ -58,7 +58,7 @@ public class UserSessionConcurrencyTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setSsoSessionIdleTimeout(1800);
         realm.setSsoSessionMaxLifespan(36000);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionExpirationTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionExpirationTest.java
@@ -40,7 +40,7 @@ public class UserSessionExpirationTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
 
         s.users().addUser(realm, "user1").setEmail("user1@localhost");

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java
@@ -67,7 +67,7 @@ public class UserSessionInitializerTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setOfflineSessionIdleTimeout(Constants.DEFAULT_OFFLINE_SESSION_IDLE_TIMEOUT);
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setSsoSessionIdleTimeout(1800);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java
@@ -77,7 +77,7 @@ public class UserSessionPersisterProviderTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setOfflineSessionIdleTimeout(Constants.DEFAULT_OFFLINE_SESSION_IDLE_TIMEOUT);
         realm.setOfflineSessionMaxLifespan(Constants.DEFAULT_OFFLINE_SESSION_MAX_LIFESPAN);
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderModelTest.java
@@ -77,7 +77,7 @@ public class UserSessionProviderModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setOfflineSessionIdleTimeout(Constants.DEFAULT_OFFLINE_SESSION_IDLE_TIMEOUT);
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setSsoSessionIdleTimeout(1800);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderOfflineModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderOfflineModelTest.java
@@ -78,7 +78,7 @@ public class UserSessionProviderOfflineModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("test");
+        RealmModel realm = createRealm(s, "test");
         realm.setOfflineSessionIdleTimeout(Constants.DEFAULT_OFFLINE_SESSION_IDLE_TIMEOUT);
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realm.setOfflineSessionMaxLifespanEnabled(true);

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/singleUseObject/SingleUseObjectModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/singleUseObject/SingleUseObjectModelTest.java
@@ -51,7 +51,7 @@ public class SingleUseObjectModelTest extends KeycloakModelTest {
 
     @Override
     public void createEnvironment(KeycloakSession s) {
-        RealmModel realm = s.realms().createRealm("realm");
+        RealmModel realm = createRealm(s, "realm");
         realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
         realmId = realm.getId();
         UserModel user = s.users().addUser(realm, "user");

